dbms/src/Common/MutexProtected.h

```diff
@@ -0,0 +1,101 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/nocopyable.h>
+#include <common/defines.h>
+#include <common/types.h>
+
+namespace DB
+{
+
+template <typename T>
+class MutexProtected
+{
+    DISALLOW_COPY_AND_MOVE(MutexProtected);
+
+private:
+    template <typename U>
+    class Locked
+    {
+        DISALLOW_COPY_AND_MOVE(Locked);
+
+    public:
+        Locked(U & value, std::mutex & mutex)
```

Comment:
Why introduce `U`, and why use `U &`?

Reply:
1. To distinct `T` and `const T`.
2. MutexProtected::Lock::value refer to MutexProtected::value

Reply:
Then can we use universal reference?

Reply:
Sorry, what does universal reference mean? Now it does refer to outer class value.

Reply:
You can check "perfect forwarding" and "universal reference". I think what you want to do is something like "reference collapsing".

Reply:
Sorry, still can not get your point, it is a lvalue reference.

Reply:
`const T &` or `T &`

Reply:
OK, I realized then...